File: 2024/Go/solutions/day10.go
package solutions

import (
    "bufio"
    "fmt"
    "log"
    "os"
)

func getTrailHeadRating(r,c int,grid[][]int) int{
    directions := [][]int{
        {1,0},
        {-1,0},
        {0,1},
        {0,-1},
    }
    if grid[r][c] == 9{
        return 1
    }
    ways := 0
    for _,direction := range(directions){
        newR := r + direction[0]
        newC := c + direction[1]
        if newR < 0 || newR >= len(grid) || newC < 0 || newC >= len(grid[0]){
            continue
        }
        if grid[newR][newC] - grid[r][c] == 1{
            ways += getTrailHeadRating(newR,newC,grid)
        }
    }
    return ways
}
func getTrailHeadScore(r,c int,grid [][]int) int{
    score := 0
    rows,cols := len(grid),len(grid[0])
    directions := [][]int{
        {1,0},
        {-1,0},
        {0,1},
        {0,-1},
    }
    queue := [][]int{}
    queue = append(queue, []int{r,c})
    queueHead := 0
    seen := make([][]int,rows,rows)
    for r := range(rows){
        seen[r] = make([]int,cols,cols)
    }
    for queueHead < len(queue){
        pos := queue[queueHead]
        queueHead += 1
        if grid[pos[0]][pos[1]] == 9{
            score += 1
            continue
        }
        for _,direction := range(directions){
            newX := pos[0] + direction[0]
            newY := pos[1] + direction[1]
            if newX < 0 || newX >= rows || newY < 0 || newY >= cols{
                continue
            }
            if grid[newX][newY] - grid[pos[0]][pos[1]] != 1{
                continue
            }
            if seen[newX][newY] == 1{
                continue
            }
            seen[newX][newY] = 1
            queue = append(queue, []int{newX,newY})
        } 
    }
    // fmt.Printf("Score from trail head at %v,%v is %v\n",r,c,score)
    return score
}

func Day10(){
    file,err := os.Open("../input.txt")
    if err != nil{
        log.Fatalf("Error while opening the file: %v",err)
    }
    part1,part2 := 0,0
    defer file.Close()
    scanner := bufio.NewScanner(file)
    grid := [][]int{}
    for scanner.Scan(){
        line := scanner.Text()
        row := []int{}
        for idx := range(len(line)){
            row = append(row, int(line[idx] - '0'))
        }
        grid = append(grid, row)
    }
    rows,cols := len(grid),len(grid[0])
    trailHeads := [][]int{}
    for r := range(rows){
        for c := range(cols){
            if grid[r][c] == 0{
                trailHeads = append(trailHeads, []int{r,c})
            }
        }
    }
    for _,trailHead := range(trailHeads){
        part1 += getTrailHeadScore(trailHead[0],trailHead[1],grid)
        part2 += getTrailHeadRating(trailHead[0],trailHead[1],grid)
    }
    fmt.Printf("Part 1: %v, Part 2 : %v\n",part1,part2)
}

File: 2024/Zig/src/day10.zig
const std = @import("std");

const directions = [4][2]i32{
    [2]i32{ 1, 0 },
    [2]i32{ -1, 0 },
    [2]i32{ 0, 1 },
    [2]i32{ 0, -1 },
};

fn getTrailHeadRating(row: usize, col: usize, grid: std.ArrayList([]u8)) u32 {
    if (grid.items[row][col] == '9') {
        return 1;
    }
    const rows: i32 = @intCast(grid.items.len);
    const cols: i32 = @intCast(grid.items[0].len);
    const r: i32 = @intCast(row);
    const c: i32 = @intCast(col);
    var ways: u32 = 0;
    for (directions) |direction| {
        const newR: i32 = r + direction[0];
        const newC: i32 = c + direction[1];
        if (newR < 0 or newR >= rows or newC < 0 or newC >= cols) {
            continue;
        }
        const diff: i32 = @subWithOverflow((grid.items[@intCast(newR)][@intCast(newC)] - '0'), (grid.items[@intCast(r)][@intCast(c)] - '0'))[0];
        if (diff != 1) {
            continue;
        }
        ways += getTrailHeadRating(@intCast(newR), @intCast(newC), grid);
    }
    return ways;
}

fn getTrailHeadScore(allocator: std.mem.Allocator, row: usize, col: usize, grid: std.ArrayList([]u8)) !u32 {
    const rows: i32 = @intCast(grid.items.len);
    const cols: i32 = @intCast(grid.items[0].len);
    var score: u32 = 0;
    var queue = std.ArrayList([2]usize).init(allocator);
    defer queue.deinit();
    var queueIdx: usize = 0;
    try queue.append([2]usize{ row, col });
    var seen = try allocator.alloc([]usize, @intCast(rows));
    for (0..@intCast(rows)) |r| {
        seen[r] = try allocator.alloc(usize, @intCast(cols));
    }
    for (0..@intCast(rows)) |r| {
        for (0..@intCast(cols)) |c| {
            seen[r][c] = 0;
        }
    }
    defer {
        for (0..@intCast(rows)) |r| {
            allocator.free(seen[r]);
        }
        allocator.free(seen);
    }
    while (queueIdx < queue.items.len) {
        const pos = queue.items[queueIdx];
        const r: i32 = @intCast(pos[0]);
        const c: i32 = @intCast(pos[1]);
        queueIdx += 1;
        if (grid.items[@intCast(r)][@intCast(c)] == '9') {
            score += 1;
            continue;
        }
        for (directions) |direction| {
            const newR: i32 = r + direction[0];
            const newC: i32 = c + direction[1];
            if (newR < 0 or newR >= rows or newC < 0 or newC >= cols) {
                continue;
            }
            const diff: i32 = @subWithOverflow((grid.items[@intCast(newR)][@intCast(newC)] - '0'), (grid.items[@intCast(r)][@intCast(c)] - '0'))[0];
            if (diff != 1) {
                continue;
            }
            if (seen[@intCast(newR)][@intCast(newC)] == 1) {
                continue;
            }
            seen[@intCast(newR)][@intCast(newC)] = 1;
            try queue.append([2]usize{ @intCast(newR), @intCast(newC) });
        }
    }
    return score;
}
fn day10_sol(allocator: std.mem.Allocator) !void {
    var part1: u32 = 0;
    var part2: u32 = 0;
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();
    var grid = std.ArrayList([]u8).init(allocator);
    defer {
        for (grid.items) |row| {
            allocator.free(row);
        }
        grid.deinit();
    }
    while (try file.reader().readUntilDelimiterOrEofAlloc(allocator, '\n', std.math.maxInt(u64))) |line| {
        try grid.append(line);
    }
    const rows: usize = grid.items.len;
    const cols: usize = grid.items[0].len;
    for (0..rows) |r| {
        for (0..cols) |c| {
            if (grid.items[r][c] == '0') {
                part1 += try getTrailHeadScore(allocator, r, c, grid);
                part2 += getTrailHeadRating(r, c, grid);
            }
        }
    }
    std.debug.print("Part 1 : {d},Part 2 : {d}\n", .{ part1, part2 });
}

test "day10" {
    try day10_sol(std.testing.allocator);
}
